File: cli/src/commands/inspect.rs
use anyhow::{anyhow, Context, Result};
use clap::Args;
use owo_colors::OwoColorize;
use serde::Deserialize;
use std::collections::BTreeMap;

use crate::logger;

#[derive(Args)]
pub struct InspectArgs {
    /// Subcommand: anr | agents | agent | capabilities
    pub subcommand: Option<String>,
    /// Target (ANR text, agent-id, etc.)
    pub target: Option<String>,
    #[arg(long, default_value = "localhost")]
    pub host: String,
    #[arg(short, long, default_value = "6174")]
    pub port: u16,
    /// Print raw JSON instead of formatted output
    #[arg(long)]
    pub raw: bool,
}

pub fn run(args: InspectArgs) -> Result<()> {
    match args.subcommand.as_deref() {
        Some("anr") => {
            let text = args
                .target
                .ok_or_else(|| anyhow!("Usage: sentrix inspect anr <anr-text>"))?;
            inspect_anr(&text, args.raw)
        }
        Some("agents") => inspect_agents(&args.host, args.port, args.raw),
        Some("agent") => {
            let id = args
                .target
                .ok_or_else(|| anyhow!("Usage: sentrix inspect agent <agent-id>"))?;
            inspect_agent(&id, &args.host, args.port, args.raw)
        }
        Some("capabilities") => inspect_capabilities(&args.host, args.port, args.raw),
        Some(other) => Err(anyhow!(
            "Unknown subcommand '{other}'. Try: anr | agents | agent | capabilities"
        )),
        None => {
            println!("\n{}", "sentrix inspect — subcommands".bold());
            println!();
            println!(
                "  {}  Decode an ANR record",
                "sentrix inspect anr <anr-text>  ".cyan()
            );
            println!(
                "  {}  List all mesh agents",
                "sentrix inspect agents           ".cyan()
            );
            println!(
                "  {}  Inspect one agent",
                "sentrix inspect agent <agent-id> ".cyan()
            );
            println!(
                "  {}  All capabilities on the mesh",
                "sentrix inspect capabilities     ".cyan()
            );
            println!();
            println!(
                "  {} --host <host>   discovery/agent host  (default: localhost)",
                "Options:".bright_black()
            );
            println!(
                "  {} --port <port>   discovery/agent port  (default: 6174)",
                "        ".bright_black()
            );
            println!(
                "  {} --raw           print raw JSON",
                "        ".bright_black()
            );
            Ok(())
        }
    }
}

// ── ANR decoder ───────────────────────────────────────────────────────────────

const ANR_PREFIX: &str = "anr:";

fn anr_key_names(key: &str) -> &str {
    match key {
        "id" => "id-scheme",
        "secp256k1" => "public-key (secp256k1)",
        "ip" => "IPv4",
        "ip6" => "IPv6",
        "tcp" => "TCP port",
        "udp" => "UDP port",
        "a.id" => "agent-id",
        "a.name" => "name",
        "a.ver" => "version",
        "a.caps" => "capabilities",
        "a.tags" => "tags",
        "a.proto" => "protocol",
        "a.port" => "agent port",
        "a.tls" => "TLS",
        "a.meta" => "metadata URI",
        "a.owner" => "owner",
        "a.chain" => "chain ID",
        other => other,
    }
}

fn inspect_anr(text: &str, raw: bool) -> Result<()> {
    let text = text.trim();
    if !text.starts_with(ANR_PREFIX) {
        return Err(anyhow!("ANR text must start with \"anr:\""));
    }
    let b64 = &text[ANR_PREFIX.len()..];

    use base64::Engine as _;
    let bytes = base64::engine::general_purpose::URL_SAFE_NO_PAD
        .decode(b64)
        .context("Failed to decode ANR base64url")?;

    let items = rlp_decode_list(&bytes).ok_or_else(|| anyhow!("ANR decode: expected RLP list"))?;

    if items.len() < 2 {
        return Err(anyhow!("ANR decode: expected at least [sig, seq, ...kv]"));
    }

    if raw {
        let hex_items: Vec<String> = items.iter().map(|b| hex(b)).collect();
        println!(
            "{}",
            serde_json::to_string_pretty(&serde_json::json!({
                "raw":   hex(&bytes),
                "items": hex_items,
            }))?
        );
        return Ok(());
    }

    println!("\n{}", "ANR Record".bold());
    println!();

    let sig_hex = hex(&items[0]);
    println!(
        "  {:<22} {}…",
        "signature".bright_black(),
        &sig_hex[..sig_hex.len().min(32)]
    );

    let seq = read_u64_be(&items[1]);
    println!("  {:<22} {seq}", "sequence".bright_black());
    println!();

    let kv = &items[2..];
    let mut i = 0;
    while i + 1 < kv.len() {
        let key_str = std::str::from_utf8(&kv[i]).unwrap_or("?").to_string();
        let label = anr_key_names(&key_str);
        let display = format_anr_value(&key_str, &kv[i + 1]);
        println!("  {:<22} {}", label.bright_black(), display);
        i += 2;
    }

    println!();
    println!(
        "  {:<22} {} / 512 bytes",
        "size".bright_black(),
        bytes.len()
    );
    if bytes.len() > 400 {
        logger::warn(&format!(
            "Record is {} bytes — approaching 512-byte limit",
            bytes.len()
        ));
    }
    Ok(())
}

fn format_anr_value(key: &str, buf: &[u8]) -> String {
    match key {
        "tcp" | "udp" | "a.port" => {
            if buf.len() >= 2 {
                let port = u16::from_be_bytes([buf[0], buf[1]]);
                port.to_string()
            } else {
                hex(buf)
            }
        }
        "a.chain" => {
            if buf.len() >= 8 {
                read_u64_be(buf).to_string()
            } else {
                hex(buf)
            }
        }
        "a.tls" => {
            if buf.first() == Some(&1) {
                "true".into()
            } else {
                "false".into()
            }
        }
        "ip" if buf.len() == 4 => {
            format!("{}.{}.{}.{}", buf[0], buf[1], buf[2], buf[3])
        }
        "secp256k1" => hex(buf),
        "a.caps" | "a.tags" => {
            if let Some(items) = rlp_decode_list(buf) {
                items
                    .iter()
                    .map(|b| String::from_utf8_lossy(b).into_owned())
                    .collect::<Vec<_>>()
                    .join(", ")
            } else {
                hex(buf)
            }
        }
        _ => {
            if let Ok(s) = std::str::from_utf8(buf) {
                if s.chars().all(|c| c.is_ascii_graphic() || c == ' ') {
                    return s.to_string();
                }
            }
            hex(buf)
        }
    }
}

// ── minimal self-contained RLP decoder ───────────────────────────────────────

/// Decode an RLP-encoded outer list, returning its elements as byte slices.
fn rlp_decode_list(data: &[u8]) -> Option<Vec<Vec<u8>>> {
    if data.is_empty() {
        return None;
    }
    let prefix = data[0] as usize;
    let (payload_start, payload_len) = if prefix <= 0xf7 {
        (1, prefix - 0xc0)
    } else {
        let ll = prefix - 0xf7;
        if data.len() < 1 + ll {
            return None;
        }
        let len = read_be_uint(&data[1..1 + ll]);
        (1 + ll, len)
    };
    if data.len() < payload_start + payload_len {
        return None;
    }

    let mut items = Vec::new();
    let mut pos = payload_start;
    let end = payload_start + payload_len;
    while pos < end {
        let (item, next) = rlp_item_at(data, pos)?;
        items.push(item.to_vec());
        pos = next;
    }
    Some(items)
}

fn rlp_item_at(data: &[u8], offset: usize) -> Option<(&[u8], usize)> {
    if offset >= data.len() {
        return None;
    }
    let prefix = data[offset] as usize;
    if prefix < 0x80 {
        return Some((&data[offset..offset + 1], offset + 1));
    }
    if prefix <= 0xb7 {
        let len = prefix - 0x80;
        let start = offset + 1;
        return Some((&data[start..start + len], start + len));
    }
    if prefix <= 0xbf {
        let ll = prefix - 0xb7;
        let len = read_be_uint(&data[offset + 1..offset + 1 + ll]);
        let start = offset + 1 + ll;
        return Some((&data[start..start + len], start + len));
    }
    // list — return entire encoded list as opaque bytes
    let (payload_start, payload_len) = if prefix <= 0xf7 {
        (offset + 1, prefix - 0xc0)
    } else {
        let ll = prefix - 0xf7;
        let len = read_be_uint(&data[offset + 1..offset + 1 + ll]);
        (offset + 1 + ll, len)
    };
    let end = payload_start + payload_len;
    Some((&data[offset..end], end))
}

fn read_be_uint(b: &[u8]) -> usize {
    b.iter().fold(0usize, |acc, &x| (acc << 8) | x as usize)
}

fn read_u64_be(b: &[u8]) -> u64 {
    b.iter().take(8).fold(0u64, |acc, &x| (acc << 8) | x as u64)
}

fn hex(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

// ── agents list ───────────────────────────────────────────────────────────────

#[derive(Deserialize, serde::Serialize, Default)]
#[serde(rename_all = "camelCase")]
struct AgentRecord {
    agent_id: String,
    #[serde(default)]
    name: String,
    #[serde(default)]
    owner: String,
    #[serde(default)]
    capabilities: Vec<String>,
    #[serde(default)]
    network: NetworkInfo,
    #[serde(default)]
    health: HealthInfo,
}

#[derive(Deserialize, serde::Serialize, Default)]
struct NetworkInfo {
    #[serde(default)]
    protocol: String,
    #[serde(default)]
    host: String,
    #[serde(default)]
    port: u16,
}

#[derive(Deserialize, serde::Serialize, Default)]
#[serde(rename_all = "camelCase")]
struct HealthInfo {
    #[serde(default)]
    status: String,
    #[serde(default)]
    last_heartbeat: String,
}

fn fetch_agents(host: &str, port: u16) -> Result<Vec<AgentRecord>> {
    let url = format!("http://{host}:{port}/agents");
    let body: serde_json::Value = reqwest::blocking::get(&url)
        .with_context(|| format!("Could not reach {url}"))?
        .json()?;
    let arr = if body.is_array() {
        body
    } else {
        body.get("agents")
            .cloned()
            .unwrap_or(serde_json::Value::Array(vec![]))
    };
    Ok(serde_json::from_value(arr)?)
}

fn health_icon(status: &str) -> &'static str {
    match status {
        "healthy" => "✔",
        "degraded" => "⚠",
        _ => "✘",
    }
}

fn inspect_agents(host: &str, port: u16, raw: bool) -> Result<()> {
    println!("\n{}", "Mesh Agents".bold());
    logger::info(&format!("Querying: http://{host}:{port}/agents"));
    println!();

    let agents = fetch_agents(host, port)?;
    if agents.is_empty() {
        logger::warn("No agents registered.");
        return Ok(());
    }

    if raw {
        println!(
            "{}",
            serde_json::to_string_pretty(&serde_json::to_value(&agents)?)?
        );
        return Ok(());
    }

    for a in &agents {
        let icon = health_icon(&a.health.status);
        println!("  {}  {}", icon.green(), a.name.bold());
        println!("  {:<20} {}", "ID:".bright_black(), a.agent_id.dimmed());
        if !a.owner.is_empty() {
            println!("  {:<20} {}", "Owner:".bright_black(), &a.owner);
        }
        if !a.capabilities.is_empty() {
            println!(
                "  {:<20} {}",
                "Capabilities:".bright_black(),
                a.capabilities.join(", ").cyan()
            );
        }
        if !a.network.host.is_empty() {
            println!(
                "  {:<20} {}://{}:{}",
                "Endpoint:".bright_black(),
                a.network.protocol,
                a.network.host,
                a.network.port
            );
        }
        if !a.health.last_heartbeat.is_empty() {
            println!(
                "  {:<20} {}",
                "Last heartbeat:".bright_black(),
                &a.health.last_heartbeat
            );
        }
        println!();
    }
    println!("  Total: {} agent(s)", agents.len().to_string().green());
    Ok(())
}

// ── single agent ─────────────────────────────────────────────────────────────

fn inspect_agent(id: &str, host: &str, port: u16, raw: bool) -> Result<()> {
    println!("\n{} {}", "Inspecting:".bold(), id.cyan());
    println!();

    // Fetch /health, /anr, /capabilities in parallel threads
    let base = format!("http://{host}:{port}");
    let (tx1, rx1) = std::sync::mpsc::channel();
    let (tx2, rx2) = std::sync::mpsc::channel();
    let (tx3, rx3) = std::sync::mpsc::channel();

    let b1 = base.clone();
    std::thread::spawn(move || {
        tx1.send(
            reqwest::blocking::get(format!("{b1}/health"))
                .and_then(|r| r.text())
                .ok(),
        )
        .ok();
    });
    let b2 = base.clone();
    std::thread::spawn(move || {
        tx2.send(
            reqwest::blocking::get(format!("{b2}/anr"))
                .and_then(|r| r.text())
                .ok(),
        )
        .ok();
    });
    let b3 = base.clone();
    std::thread::spawn(move || {
        tx3.send(
            reqwest::blocking::get(format!("{b3}/capabilities"))
                .and_then(|r| r.text())
                .ok(),
        )
        .ok();
    });

    let health_raw = rx1.recv().ok().flatten();
    let anr_raw = rx2.recv().ok().flatten();
    let caps_raw = rx3.recv().ok().flatten();

    if raw {
        let parse =
            |s: Option<String>| s.and_then(|t| serde_json::from_str::<serde_json::Value>(&t).ok());
        println!(
            "{}",
            serde_json::to_string_pretty(&serde_json::json!({
                "health":       parse(health_raw),
                "anr":          parse(anr_raw),
                "capabilities": parse(caps_raw),
            }))?
        );
        return Ok(());
    }

    if let Some(h) = health_raw
        .as_deref()
        .and_then(|s| serde_json::from_str::<serde_json::Value>(s).ok())
    {
        println!(
            "  {:<22} {}",
            "Status:".bright_black(),
            h["status"].as_str().unwrap_or("—")
        );
        println!(
            "  {:<22} {}",
            "Agent ID:".bright_black(),
            h["agentId"].as_str().unwrap_or("—")
        );
        if let Some(v) = h["version"].as_str() {
            println!("  {:<22} {}", "Version:".bright_black(), v);
        }
        if let Some(u) = h["uptimeMs"].as_u64() {
            println!("  {:<22} {}s", "Uptime:".bright_black(), u / 1000);
        }
        if let Some(n) = h["capabilitiesCount"].as_u64() {
            println!("  {:<22} {n}", "Capabilities:".bright_black());
        }
    } else {
        logger::warn(&format!(
            "Health endpoint unreachable at http://{host}:{port}/health"
        ));
    }
    println!();

    if let Some(c) = caps_raw
        .as_deref()
        .and_then(|s| serde_json::from_str::<serde_json::Value>(s).ok())
    {
        if let Some(caps) = c["capabilities"].as_array() {
            println!("  {}", "Capabilities:".bold());
            for cap in caps {
                let name = cap["name"].as_str().unwrap_or(cap.as_str().unwrap_or("?"));
                let price = cap["price"]
                    .as_str()
                    .map(|p| format!("  [{}]", p))
                    .unwrap_or_default();
                println!("    • {}{}", name.cyan(), price);
                if let Some(d) = cap["description"].as_str() {
                    println!("      {}", d.dimmed());
                }
            }
            println!();
        }
    }

    if let Some(a) = anr_raw
        .as_deref()
        .and_then(|s| serde_json::from_str::<serde_json::Value>(s).ok())
    {
        if let Some(anr_text) = a["anr"].as_str() {
            println!("  {}", "ANR:".bold());
            println!("    {}", anr_text.dimmed());
            println!();
            println!("  Decode with: sentrix inspect anr \"{}\"", anr_text);
        }
    }
    Ok(())
}

// ── capabilities map ──────────────────────────────────────────────────────────

fn inspect_capabilities(host: &str, port: u16, raw: bool) -> Result<()> {
    println!("\n{}", "All Mesh Capabilities".bold());
    logger::info(&format!("Querying: http://{host}:{port}/agents"));
    println!();

    let agents = fetch_agents(host, port)?;
    let mut cap_map: BTreeMap<String, Vec<String>> = BTreeMap::new();

    for a in &agents {
        for cap in &a.capabilities {
            cap_map
                .entry(cap.clone())
                .or_default()
                .push(if a.name.is_empty() {
                    a.agent_id.clone()
                } else {
                    a.name.clone()
                });
        }
    }

    if cap_map.is_empty() {
        logger::warn("No capabilities found.");
        return Ok(());
    }

    if raw {
        println!("{}", serde_json::to_string_pretty(&cap_map)?);
        return Ok(());
    }

    for (cap, providers) in &cap_map {
        let reserved = if cap.starts_with("__") {
            " (reserved)".dimmed().to_string()
        } else {
            String::new()
        };
        println!("  {}{}", cap.cyan(), reserved);
        for p in providers {
            println!("    {} {}", "↳".bright_black(), p);
        }
    }

    println!();
    println!(
        "  {} unique capability(-ies) across {} agent(s)",
        cap_map.len().to_string().green(),
        agents.len().to_string().green()
    );
    Ok(())
}

File: src/commands/inspect.ts
/**
 * borgkit inspect — inspect ANR records and mesh topology.
 *
 * Subcommands:
 *   borgkit inspect anr <anr-text>            Decode and display an ANR record
 *   borgkit inspect agents [--host] [--port]  List all agents in the mesh
 *   borgkit inspect agent <agent-id>          Inspect a specific agent
 *   borgkit inspect capabilities              List all capabilities in the mesh
 */

import http from 'http';
import { logger } from '../utils/logger';

// ── ANR constants (mirrors templates/typescript/anr/anr.ts) ──────────────────

const ANR_PREFIX    = 'anr:';
const ANR_KEY_NAMES: Record<string, string> = {
  'id':        'id-scheme',
  'secp256k1': 'public-key (secp256k1)',
  'ip':        'IPv4',
  'ip6':       'IPv6',
  'tcp':       'TCP port',
  'udp':       'UDP port',
  'a.id':      'agent-id',
  'a.name':    'name',
  'a.ver':     'version',
  'a.caps':    'capabilities (RLP)',
  'a.tags':    'tags (RLP)',
  'a.proto':   'protocol',
  'a.port':    'agent port',
  'a.tls':     'TLS',
  'a.meta':    'metadata URI',
  'a.owner':   'owner',
  'a.chain':   'chain ID',
};

// ── subcommand dispatcher ─────────────────────────────────────────────────────

export interface InspectOptions {
  host?:       string;
  port?:       string;
  capability?: string;
  raw?:        boolean;
}

export async function inspectCommand(
  subcommand: string | undefined,
  target:     string | undefined,
  options:    InspectOptions,
): Promise<void> {
  switch (subcommand) {
    case 'anr':
      if (!target) {
        logger.error('Usage: borgkit inspect anr <anr-text>');
        process.exit(1);
      }
      return inspectAnr(target, options);

    case 'agents':
      return inspectAgents(options);

    case 'agent':
      if (!target) {
        logger.error('Usage: borgkit inspect agent <agent-id>');
        process.exit(1);
      }
      return inspectAgent(target, options);

    case 'capabilities':
      return inspectCapabilities(options);

    default:
      logger.title('borgkit inspect — available subcommands');
      console.log('');
      console.log('  borgkit inspect anr <anr-text>           Decode an ANR record');
      console.log('  borgkit inspect agents                   List all mesh agents');
      console.log('  borgkit inspect agent <agent-id>         Inspect one agent');
      console.log('  borgkit inspect capabilities             All mesh capabilities');
      console.log('');
      console.log('Options:');
      console.log('  --host <host>   Discovery/agent host  (default: localhost)');
      console.log('  --port <port>   Discovery/agent port  (default: 3000 for agents, 8080 for agent)');
      console.log('  --raw           Print raw JSON');
      console.log('');
  }
}

// ── ANR decoder ───────────────────────────────────────────────────────────────

function inspectAnr(anrText: string, options: InspectOptions): void {
  if (!anrText.startsWith(ANR_PREFIX)) {
    logger.error(`ANR text must start with "${ANR_PREFIX}"`);
    process.exit(1);
  }

  let raw: Buffer;
  try {
    const b64 = anrText.slice(ANR_PREFIX.length).replace(/-/g, '+').replace(/_/g, '/');
    const padded = b64 + '='.repeat((4 - b64.length % 4) % 4);
    raw = Buffer.from(padded, 'base64');
  } catch {
    logger.error('Failed to decode ANR base64');
    process.exit(1);
  }

  try {
    const decoded = rlpDecode(raw);

    if (options.raw) {
      // Print as hex key-value pairs
      console.log(JSON.stringify(
        { raw: raw.toString('hex'), decoded: formatRlpForJson(decoded) },
        null, 2,
      ));
      return;
    }

    if (!Array.isArray(decoded) || decoded.length < 2) {
      logger.error('ANR decode: expected RLP list with at least [sig, seq, ...kv]');
      process.exit(1);
    }

    const [sigBuf, seqBuf, ...rest] = decoded as Buffer[];

    logger.title('ANR Record');
    console.log('');

    // Signature
    const sigHex = Buffer.isBuffer(sigBuf) ? sigBuf.toString('hex') : '?';
    console.log(`  ${'signature'.padEnd(22)} ${sigHex.slice(0, 32)}...`);

    // Sequence
    const seq = Buffer.isBuffer(seqBuf) ? readUint64BE(seqBuf) : 0;
    console.log(`  ${'sequence'.padEnd(22)} ${seq}`);
    console.log('');

    // Key-value pairs
    for (let i = 0; i + 1 < rest.length; i += 2) {
      const keyBuf = rest[i];
      const valBuf = rest[i + 1];
      if (!Buffer.isBuffer(keyBuf) || !Buffer.isBuffer(valBuf)) continue;

      const key     = keyBuf.toString('utf8');
      const label   = ANR_KEY_NAMES[key] ?? key;
      const display = formatAnrValue(key, valBuf);
      console.log(`  ${label.padEnd(22)} ${display}`);
    }

    console.log('');
    console.log(`  ${'size'.padEnd(22)} ${raw.length} / 512 bytes`);
    if (raw.length > 400) {
      logger.warn(`  Record is ${raw.length} bytes — approaching 512-byte limit`);
    }
  } catch (e) {
    logger.error(`ANR decode failed: ${e}`);
    process.exit(1);
  }
}

function formatAnrValue(key: string, buf: Buffer): string {
  // Port fields (2-byte big-endian uint16)
  if (['tcp', 'udp', 'a.port'].includes(key)) {
    return buf.length >= 2 ? buf.readUInt16BE(0).toString() : buf.toString('hex');
  }
  // Chain ID (8-byte big-endian uint64)
  if (key === 'a.chain') {
    return buf.length >= 8 ? readUint64BE(buf).toString() : buf.toString('hex');
  }
  // Boolean (TLS)
  if (key === 'a.tls') {
    return buf.length > 0 ? (buf[0] === 1 ? 'true' : 'false') : 'false';
  }
  // IPv4 (4 bytes)
  if (key === 'ip' && buf.length === 4) {
    return Array.from(buf).join('.');
  }
  // Public key — show compressed hex
  if (key === 'secp256k1') {
    return buf.toString('hex');
  }
  // Capabilities / tags — RLP list
  if (key === 'a.caps' || key === 'a.tags') {
    try {
      const items = rlpDecode(buf) as Buffer[];
      if (Array.isArray(items)) {
        return items.map(b => b.toString('utf8')).join(', ');
      }
    } catch { /* fall through */ }
  }
  // Default: try UTF-8, fallback to hex
  try {
    const str = buf.toString('utf8');
    if (/^[\x20-\x7E]*$/.test(str)) return str;
  } catch { /* */ }
  return buf.toString('hex');
}

// ── agents list ───────────────────────────────────────────────────────────────

async function inspectAgents(options: InspectOptions): Promise<void> {
  const host = options.host ?? 'localhost';
  const port = options.port ?? '3000';
  const url  = `http://${host}:${port}/agents`;

  logger.title('Mesh Agents');
  logger.info(`Querying: ${url}`);
  console.log('');

  let agents: AgentRecord[];
  try {
    agents = JSON.parse(await httpGet(url));
  } catch (e) {
    logger.error(`Failed to reach discovery layer: ${e}`);
    logger.dim(`Is a discovery server running at ${host}:${port}?`);
    process.exit(1);
  }

  if (!agents.length) {
    logger.warn('No agents registered in the discovery layer.');
    return;
  }

  if (options.raw) {
    console.log(JSON.stringify(agents, null, 2));
    return;
  }

  const statusIcon = (s: string) => s === 'healthy' ? '✔' : s === 'degraded' ? '⚠' : '✘';

  for (const a of agents) {
    const icon = statusIcon(a.health?.status ?? 'unknown');
    console.log(`  ${icon}  ${a.name ?? a.agentId}`);
    console.log(`     ID:           ${a.agentId}`);
    console.log(`     Owner:        ${a.owner}`);
    console.log(`     Capabilities: ${(a.capabilities ?? []).join(', ')}`);
    console.log(`     Endpoint:     ${a.network?.protocol ?? 'http'}://${a.network?.host}:${a.network?.port}`);
    console.log(`     Health:       ${a.health?.status ?? 'unknown'}`);
    if (a.health?.lastHeartbeat) {
      const ago = Math.round((Date.now() - new Date(a.health.lastHeartbeat).getTime()) / 1000);
      console.log(`     Last seen:    ${ago}s ago`);
    }
    console.log('');
  }

  console.log(`  Total: ${agents.length} agent(s)`);
}

// ── single agent inspect ──────────────────────────────────────────────────────

async function inspectAgent(agentId: string, options: InspectOptions): Promise<void> {
  const host = options.host ?? 'localhost';
  const port = options.port ?? '8080';

  logger.title(`Inspecting agent: ${agentId}`);
  console.log('');

  // Fetch health, ANR, and capabilities in parallel
  const [healthRaw, anrRaw, capsRaw] = await Promise.allSettled([
    httpGet(`http://${host}:${port}/health`),
    httpGet(`http://${host}:${port}/anr`),
    httpGet(`http://${host}:${port}/capabilities`),
  ]);

  if (options.raw) {
    console.log(JSON.stringify({
      health:       healthRaw.status === 'fulfilled' ? JSON.parse(healthRaw.value) : null,
      anr:          anrRaw.status    === 'fulfilled' ? JSON.parse(anrRaw.value)    : null,
      capabilities: capsRaw.status   === 'fulfilled' ? JSON.parse(capsRaw.value)   : null,
    }, null, 2));
    return;
  }

  if (healthRaw.status === 'fulfilled') {
    try {
      const h = JSON.parse(healthRaw.value) as HealthResponse;
      console.log(`  Status:       ${h.status}`);
      console.log(`  Agent ID:     ${h.agentId ?? '—'}`);
      console.log(`  Version:      ${h.version ?? '—'}`);
      console.log(`  Uptime:       ${h.uptimeMs != null ? `${Math.round(h.uptimeMs / 1000)}s` : '—'}`);
      console.log(`  Capabilities: ${h.capabilitiesCount ?? '—'}`);
    } catch { /* */ }
  } else {
    logger.warn(`Health endpoint unreachable at http://${host}:${port}/health`);
  }

  console.log('');

  if (capsRaw.status === 'fulfilled') {
    try {
      const caps = JSON.parse(capsRaw.value) as { capabilities: CapSummary[] };
      console.log('  Capabilities:');
      for (const cap of caps.capabilities ?? []) {
        const price = cap.price ? `  [${cap.price}]` : '';
        console.log(`    • ${cap.name}${price}`);
        if (cap.description) console.log(`      ${cap.description}`);
      }
    } catch { /* */ }
  }

  console.log('');

  if (anrRaw.status === 'fulfilled') {
    try {
      const anr = JSON.parse(anrRaw.value);
      if (anr.anr) {
        console.log('  ANR text:');
        console.log(`    ${anr.anr}`);
        console.log('');
        console.log(`  Decode with: borgkit inspect anr "${anr.anr}"`);
      }
    } catch { /* */ }
  }
}

// ── capabilities list ─────────────────────────────────────────────────────────

async function inspectCapabilities(options: InspectOptions): Promise<void> {
  const host = options.host ?? 'localhost';
  const port = options.port ?? '3000';
  const url  = `http://${host}:${port}/agents`;

  logger.title('All Mesh Capabilities');
  logger.info(`Querying: ${url}`);
  console.log('');

  let agents: AgentRecord[];
  try {
    agents = JSON.parse(await httpGet(url));
  } catch (e) {
    logger.error(`Failed to reach discovery layer: ${e}`);
    process.exit(1);
  }

  // Build capability → agents map
  const capMap = new Map<string, string[]>();
  for (const a of agents) {
    for (const cap of a.capabilities ?? []) {
      if (!capMap.has(cap)) capMap.set(cap, []);
      capMap.get(cap)!.push(a.name ?? a.agentId);
    }
  }

  if (capMap.size === 0) {
    logger.warn('No capabilities found.');
    return;
  }

  if (options.raw) {
    const out: Record<string, string[]> = {};
    for (const [cap, providers] of capMap) out[cap] = providers;
    console.log(JSON.stringify(out, null, 2));
    return;
  }

  const sorted = Array.from(capMap.entries()).sort(([a], [b]) => a.localeCompare(b));
  for (const [cap, providers] of sorted) {
    const reserved = cap.startsWith('__') ? ' (reserved)' : '';
    console.log(`  ${cap}${reserved}`);
    for (const p of providers) console.log(`    ↳ ${p}`);
  }

  console.log('');
  console.log(`  Total: ${capMap.size} unique capability(-ies) across ${agents.length} agent(s)`);
}

// ── minimal RLP decoder ───────────────────────────────────────────────────────

function rlpDecode(data: Buffer): unknown {
  const [result] = rlpDecodeAt(data, 0);
  return result;
}

function rlpDecodeAt(data: Buffer, offset: number): [unknown, number] {
  const prefix = data[offset];
  if (prefix < 0x80) {
    return [data.subarray(offset, offset + 1), offset + 1];
  }
  if (prefix <= 0xb7) {
    const len   = prefix - 0x80;
    const start = offset + 1;
    return [data.subarray(start, start + len), start + len];
  }
  if (prefix <= 0xbf) {
    const lenLen = prefix - 0xb7;
    const len    = readBigEndian(data, offset + 1, lenLen);
    const start  = offset + 1 + lenLen;
    return [data.subarray(start, start + len), start + len];
  }
  if (prefix <= 0xf7) {
    const payloadLen = prefix - 0xc0;
    const items: unknown[] = [];
    let   pos   = offset + 1;
    const end   = pos + payloadLen;
    while (pos < end) {
      const [item, next] = rlpDecodeAt(data, pos);
      items.push(item);
      pos = next;
    }
    return [items, end];
  }
  const lenLen     = prefix - 0xf7;
  const payloadLen = readBigEndian(data, offset + 1, lenLen);
  const items: unknown[] = [];
  let   pos = offset + 1 + lenLen;
  const end = pos + payloadLen;
  while (pos < end) {
    const [item, next] = rlpDecodeAt(data, pos);
    items.push(item);
    pos = next;
  }
  return [items, end];
}

function readBigEndian(buf: Buffer, offset: number, len: number): number {
  let n = 0;
  for (let i = 0; i < len; i++) n = (n * 256) + buf[offset + i];
  return n;
}

function readUint64BE(buf: Buffer): number {
  if (buf.length < 8) return 0;
  // JS can handle up to 2^53 safely
  return buf.readUInt32BE(0) * 0x100000000 + buf.readUInt32BE(4);
}

function formatRlpForJson(v: unknown): unknown {
  if (Buffer.isBuffer(v)) return v.toString('hex');
  if (Array.isArray(v))   return v.map(formatRlpForJson);
  return v;
}

// ── HTTP helper ───────────────────────────────────────────────────────────────

function httpGet(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    http.get(url, (res) => {
      let d = '';
      res.on('data', c => { d += c; });
      res.on('end', () => resolve(d));
    }).on('error', reject);
  });
}

// ── types ─────────────────────────────────────────────────────────────────────

interface AgentRecord {
  agentId:      string;
  name?:        string;
  owner:        string;
  capabilities: string[];
  network?:     { protocol: string; host: string; port: number };
  health?:      { status: string; lastHeartbeat?: string };
}

interface HealthResponse {
  status:            string;
  agentId?:          string;
  version?:          string;
  uptimeMs?:         number;
  capabilitiesCount?: number;
}

interface CapSummary {
  name:         string;
  description?: string;
  price?:       string;
}
